File: src/gui-tesing-library/Components/GTWindowCS.cs
using System;
using gui_tesing_library_CS.Directives;
using gui_tesing_library_CS.SystemCalls;
using gui_tesing_library;
using gui_tesing_library.Controllers;
using gui_tesing_library.Interfaces;
using gui_tesing_library.Models;

namespace gui_tesing_library_CS.Components;

public class GTWindowCS : IGTWindow
{
    private readonly int _handle;
    private readonly ISystemCalls _SystemCalls = SystemCallsFactory.GetSystemCalls();

    public GTWindowCS(int handle)
    {
        if (handle == 0)
            throw new ArgumentException("Window cannot have handle 0");
        _handle = handle;
    }

    public Vector2i MaximizedWindowSize { get; }

    public Vector2i Position => _SystemCalls.GetWindowPositon(_handle);

    public bool DoesExist => _SystemCalls.DoesWindowExist(_handle);

    public string Name => throw new NotImplementedException();
    public bool IsMinimized => _SystemCalls.IsWindowsMinimized(_handle);

    public Vector2i Size => _SystemCalls.GetWindowSize(_handle);

    [Delay]
    [Log]
    IGTWindow IGTWindow.Close()
    {
        _SystemCalls.CloseWindow(_handle);
        return this;
    }

    [Log]
    [Delay]
    public IGTWindow SetWindowSize(int x, int y)
    {
        _SystemCalls.SetWindowSize(_handle, new Vector2i(x, y));
        return this;
    }

    [Log]
    [Delay]
    public IGTWindow SetPostion(int x, int y)
    {
        _SystemCalls.SetWindowPostion(_handle, new Vector2i(x, y));
        return this;
    }

    [Log]
    [Delay]
    public IGTWindow BringUpFront()
    {
        _SystemCalls.BringWindowUpFront(_handle);
        return this;
    }

    [Log]
    public Color GetPixelColorAt(Vector2i postion)
    {
        return SystemCallsFactory.GetSystemCalls().GetPixelColorAt(postion, _handle);
    }

    [Log]
    public IGTScreen PixelAtShouldBeColor(Vector2i position, Color color)
    {
        Helpers.AwaitTrue(
            () => { return GetPixelColorAt(position).Equals(color); },
            $"Pixel color at {position} of window {GetWindowName()} was not {color} with in given time"
        );
        return this;
    }

    [Log]
    [Delay]
    IGTWindow IGTWindow.Maximize()
    {
        _SystemCalls.MaximizeWindow(_handle);
        return this;
    }

    [Log]
    public IGTProcess GetProcessOfWindow()
    {
        throw new NotImplementedException();
        //int processHandle = this._SystemCalls.GetProcessHandleFromId(
        //    this._SystemCalls.GetWindowThreadProcessId(this._handle)
        //);
        //return new GTProcess_CS(processHandle);
    }

    [Log]
    public IScreenShot GetScreenshot()
    {
        return SystemCallsFactory
            .GetSystemCalls()
            .GetScreenShotFromHandle(
                _handle,
                new Vector2i(0, 0),
                GetWindowContentSize()
            );
    }

    [Log]
    [Delay]
    public IScreenShot GetScreenshotRect(Vector2i position, Vector2i size)
    {
        throw new NotImplementedException();
    }

    [Log]
    [Delay]
    IGTWindow IGTWindow.Minimize()
    {
        _SystemCalls.MinimizeWindow(_handle);
        return this;
    }

    [Log]
    [Delay]
    IGTWindow IGTWindow.MoveWindow(Vector2i offset)
    {
        throw new NotImplementedException();
    }

    [Log]
    IGTWindow IGTWindow.SizeShouldBe(Vector2i vector2I)
    {
        Helpers.AwaitTrue(() => { return Size.Equals(vector2I); });
        return this;
    }

    [Log]
    IGTWindow IGTWindow.ShouldWindowExist(bool v)
    {
        Helpers.AwaitTrue(
            () => DoesExist == v,
            $"Window {GetWindowName()} is not {v} within max time"
        );
        return this;
    }

    [Log]
    IGTWindow IGTWindow.KillProcess()
    {
        GetProcessOfWindow().kill();
        return this;
    }

    [Log]
    public IGTWindow ShouldBeMinimized(bool state)
    {
        Helpers.AwaitTrue(() => { return IsMinimized == state; });
        return this;
    }

    [Log]
    public string GetWindowName()
    {
        return SystemCallsFactory.GetSystemCalls().GetWindowName(_handle);
    }

    public Vector2i GetWindowContentPosition()
    {
        var positon = Position;
        return new Vector2i(
            positon.x + SystemController.Instance.GetWindowBorderWidth(),
            positon.y
            + SystemController.Instance.GetWindowBorderHeight()
            + SystemController.Instance.GetWindowTitleBarHeight()
        );
    }

    public Vector2i GetWindowContentSize()
    {
        var size = Size;
        var borderWidth = SystemController.Instance.GetWindowBorderWidth();
        var windwopadding = SystemController.Instance.GetWindowPadding();
        var WindowBorderHeight = SystemController.Instance.GetWindowBorderHeight();
        var windowtielebarheight = SystemController.Instance.GetWindowTitleBarHeight();
        return new Vector2i(
            size.x - borderWidth * 2 - windwopadding * 2 - 1,
            size.y - WindowBorderHeight * 2 - windowtielebarheight - windwopadding * 2 - 5
        );
    }

    [Log]
    [Delay]
    public Color GetContentPixelColorAt(Vector2i postion)
    {
        //Vector2i WindowPOsiton = this.Position;
        //Vector2i ContnetPostion = this.GetWindowContentPosition();
        //Vector2i offset = ContnetPostion - WindowPOsiton;
        //Vector2i newPos = postion + offset;
        return GetPixelColorAt(postion);
    }

    public Color GetContentPixelColorAt(Vector2f realtivePostion)
    {
        var contentSize = GetWindowContentSize();
        return GetContentPixelColorAt(
            new Vector2i(
                (int)(contentSize.x * realtivePostion.x),
                (int)(contentSize.y * realtivePostion.y)
            )
        );
    }

    public IGTWindow ContentPixelAtShouldBeColor(Vector2i position, Color color)
    {
        Helpers.AwaitTrue(
            () => { return GetContentPixelColorAt(position).Equals(color); },
            $"Content Pixel color at {position} of window {GetWindowName()} was not {color} but {GetContentPixelColorAt(position)} with in given time"
        );
        return this;
    }

    public IGTWindow CenterWindow()
    {
        var screenSize = SystemController.Instance.GetScreenSize();
        var windowSize = Size;
        var diffrence = screenSize - windowSize;
        SetPostion(diffrence.x / 2, diffrence.y / 2);

        return this;
    }

    public IGTWindow WindowNameShouldBe(string title)
    {
        Helpers.AwaitTrue(
            () => { return GetWindowName() == title; },
            $"Window name was not [[{title}]] but [[{GetWindowName()}]] within {Configuration.ProcesAwaitTime} ms"
        );
        return this;
    }

    public IGTWindow ContentPixelAtShouldBeColor(
        Vector2f sliderColorCheckPostion,
        Color colorshouldbe,
        int errorPass
    )
    {
        Helpers.AwaitTrue(
            () =>
            {
                return GetContentPixelColorAt(sliderColorCheckPostion)
                    .getDiffrence(colorshouldbe) < errorPass;
            },
            $"Content Pixel color at {sliderColorCheckPostion} of window {GetWindowName()} was not {colorshouldbe} but {GetContentPixelColorAt(sliderColorCheckPostion)} with in given time, difreance: [[{GetContentPixelColorAt(sliderColorCheckPostion).getDiffrence(colorshouldbe)}]], with error pass [[{errorPass}]]"
        );
        return this;
    }

    [Log]
    [Delay]
    public IGTProcess Close()
    {
        throw new NotImplementedException();
    }

    [Log]
    [Delay]
    public IGTProcess Maximize()
    {
        throw new NotImplementedException();
    }

    [Log]
    [Delay]
    public IGTProcess Minimize()
    {
        throw new NotImplementedException();
    }

    [Log]
    [Delay]
    public IGTProcess MoveWindow(Vector2i offset)
    {
        throw new NotImplementedException();
    }
}

File: src/GTLibrary_CSharp/GTWindow_CS.cpp
#include "GTWindow_CS.h"
#include "GTWindow_CS.h"
#include "Vector2i_CS.h"
#include "Vector2f_CS.h"
#include "GTProcess_CS.h"
#include "Color_CS.h"
#include "Converters.h"
#include "GTWindow_CS.h"

Vector2i_CS^ GTWindow_CS::GetPosition() {
	return Converters::Vector2iToVector2iCS(_nativeWindow->GetPosition());
}
bool GTWindow_CS::DoesExist() {

	return this->_nativeWindow->DoesExist();
}
String^ GTWindow_CS::GetName() {
	return Converters::ConvertWStdStringToString(_nativeWindow->GetName());
}
bool GTWindow_CS::IsMinimized() {
	return _nativeWindow->IsMinimized();
}

void GTWindow_CS::MoveWindow(Vector2i_CS^ offset) {
	_nativeWindow->MoveWindow(Converters::Vector2iCSToVector2i(offset));
}
void GTWindow_CS::Minimize() {
	_nativeWindow->Minimize();
;}
void GTWindow_CS::Maximize() {
	_nativeWindow->Maximize();
}
GTProcess_CS^ GTWindow_CS::GetProcessOfWindow() {
	return Converters::ProcessToProcessCS(_nativeWindow->GetProcessOfWindow());
}
void GTWindow_CS::Close() {
	_nativeWindow->Close();
}
void GTWindow_CS::SetWindowSize(int x, int y) {
	_nativeWindow->SetWindowSize(x,y);

}
void GTWindow_CS::SetPosition(int x, int y) {
	_nativeWindow->SetPosition(x, y);
}
void GTWindow_CS::BringUpFront() {
	_nativeWindow->BringUpFront();

}
void GTWindow_CS::SizeShouldBe(Vector2i_CS^ size)
{
	_nativeWindow->SizeShouldBe(Converters::Vector2iCSToVector2i(size));
}
void GTWindow_CS::ShouldWindowExist(bool exists) {
	_nativeWindow->ShouldWindowExist(exists);
}
void GTWindow_CS::KillProcess() {
	_nativeWindow->KillProcess();

}
void GTWindow_CS::ShouldBeMinimized(bool state) {
	_nativeWindow->ShouldBeMinimized(state);
}
String^ GTWindow_CS::GetWindowName() {
	return Converters::ConvertWStdStringToString(_nativeWindow->GetName());
}
Vector2i_CS^ GTWindow_CS::GetWindowContentPosition() {
	return Converters::Vector2iToVector2iCS(_nativeWindow->GetWindowContentPosition());
}
Vector2i_CS^ GTWindow_CS::GetWindowContentSize() {
	Vector2i size = _nativeWindow->GetWindowContentSize();
	return Converters::Vector2iToVector2iCS(size);

}
Color_CS^ GTWindow_CS::GetContentPixelColorAt(Vector2i_CS^ position) {
	return Converters::ColorToColorCS(_nativeWindow->GetContentPixelColorAt(Converters::Vector2iCSToVector2i(position)));
}
Color_CS^ GTWindow_CS::GetContentPixelColorAt(Vector2f_CS^ relativePosition) {
	return Converters::ColorToColorCS(_nativeWindow->GetContentPixelColorAt(Converters::Vector2fCSToVector2f(relativePosition)));
}
void GTWindow_CS::ContentPixelAtShouldBeColor(Vector2i_CS^ position, Color_CS^ color) {
	_nativeWindow->ContentPixelAtShouldBeColor(Converters::Vector2iCSToVector2i(position), Converters::ColorCSToColor(color));

}
void GTWindow_CS::CenterWindow() {
	_nativeWindow->CenterWindow();
}
void GTWindow_CS::WindowNameShouldBe(String^ title) {
	_nativeWindow->WindowNameShouldBe(Converters::ConvertStringToWStdString(title));
}
void GTWindow_CS::ContentPixelAtShouldBeColor(Vector2f_CS^ sliderColorCheckPosition, Color_CS^ colorShouldBe, int errorPass) {

}
Vector2i_CS^ GTWindow_CS::GetWindowSize() {
	return gcnew Vector2i_CS(_nativeWindow->GetSize());
}
Vector2i_CS^ GTWindow_CS::GetMaximizedWindowSize() {
	return Converters::Vector2iToVector2iCS(GTSystem::Instance()-> GetMaximizedWindowSize());
}
ScreenShot_CS^ GTWindow_CS::GetScreenshot() {
	return Converters::ScreenShotTOScreenShotCs(_nativeWindow->GetScreenshot());
}

ScreenShot_CS^ GTWindow_CS::GetScreenshotRect(Vector2i_CS^ position, Vector2i_CS^ size) {
	return Converters::ScreenShotTOScreenShotCs(_nativeWindow->GetScreenshotRect(
		Converters::Vector2iCSToVector2i(position), Converters::Vector2iCSToVector2i(size)
	));
}

Color_CS^ GTWindow_CS::GetPixelColorAt(Vector2i_CS^ position) {
	return Converters::ColorToColorCS(_nativeWindow->GetPixelColorAt(Converters::Vector2iCSToVector2i(position)));
}

void GTWindow_CS::PixelAtShouldBeColor( Vector2i_CS^ position, Color_CS^ color) {
	_nativeWindow->PixelAtShouldBeColor(Converters::Vector2iCSToVector2i(position), Converters::ColorCSToColor(color));
}
Vector2i_CS^ GTWindow_CS::GetWindowPosition() {
	return Converters::Vector2iToVector2iCS(_nativeWindow->GetPosition());
}
GTWindow* GTWindow_CS::GetNative()
{
	return this->_nativeWindow;
}

